### PR TITLE
Use modified TimerSet in task utils with invariant

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
Ensures to-be-resumed tasks in currently registered events are in the
valid state.

### CODE REVIEW

File: src/ocean/task/util/Timer.d [hunk 1]
```diff
@@ -112,10 +112,7 @@ unittest
     theScheduler.schedule(task);
     theScheduler.eventLoop();
 
-    // NB: allocated event count is expected to be 1 more than strictly
-    // necessary here because they are recycled only after task finishes
-    // or suspend again, not immediately after it gets resumed on timer
-    test!("==")(.timer.allocated_event_count(), 2);
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/util/Timer.d:118]: Ha, nice!

File: src/ocean/io/select/client/TimerSet.d [hunk 2]
```diff
@@ -256,15 +256,7 @@ public class TimerSet ( EventData ) : TimerEventTimeoutManager
...
-    private ObjectPool!(Event) events;
-
-    version (UnitTest)
-    {
-        public size_t allocated_event_count ( )
-        {
-            return events.length();
-        }
-    }
+    protected ObjectPool!(Event) events;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/io/select/client/TimerSet.d:259]: Changing this field from private to protected seems unrelated to "Remove unnecessary unittest function".

Comment (mihails-strasuns-sociomantic) [reply src/ocean/io/select/client/TimerSet.d:259]: Good catch, moved to the next commit

**Comment by mihails-strasuns-sociomantic** [comment]
Updated:
- move protection attribute change to commit that requires it

**Comment by gavin-norman-sociomantic** [comment]
Hmm red now.

**Comment by mihails-strasuns-sociomantic** [comment]
Something weird and scary is being hit here. Can reproduce it locally time to time (but never in gdb), and the written core shows borken stack on fiber context switch.

**Comment by gavin-norman-sociomantic** [comment]
:( "Oh no, not again."

**Comment by mihails-strasuns-sociomantic** [comment]
Blocked on merge of custom pool PR (and its refactorings).

**Comment by mihails-strasuns-sociomantic** [comment]
Should be good to go now

File: src/ocean/task/util/Timer.d [hunk 3]
```diff
@@ -262,6 +262,26 @@ private struct EventData
...
+    Derivative from generic ocean TimerSet which augments it with task-specific
+    sanity checks.
+
+*******************************************************************************/
+
+private class TaskTimerSet : TimerSet!(EventData)
+{
+    invariant ( )
+    {
+        auto _this = cast(TaskTimerSet) this;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/util/Timer.d:274]: What's this cast for btw?

Comment (gavin-norman-sociomantic) [reply src/ocean/task/util/Timer.d:274]: Ahh is it `const`-related?

**Review by gavin-norman-sociomantic — APPROVED** [review]

File: src/ocean/task/util/Timer.d (hunk 3 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/util/Timer.d:274]: Yep, `this` is const in invariants
